docs/architecture/adr-042-blockchain-riri-org.md

```diff
@@ -0,0 +1,292 @@
+                schedule.stop()
+                break;
+        }
+    }
+}
+```
+
+## Lifecycle management
+
+A set of routines for individual processes allow processes to run in parallel with clear lifecycle management. `Start`, `Stop`, and `AddPeer` hooks currently present in the reactor will delegate to the sub-routines allowing them to manage internal state independant without further coupling to the reactor.
+
+```go
+func (r *BlockChainReactor) Start() {
+    r.msgs := make(chan Message, maxInFlight)
+    schedulerMsgs := make(chan Message)
+    processorMsgs := make(chan Message)
+    ioMsgs := make(chan Message)
+
+    go processorRoutine(processorMsgs, r.msgs)
+    go scheduleRoutine(schedulerMsgs, r.msgs)
+    go ioRoutine(ioMsgs, r.msgs)
+    ...
+}
+
+...
+func (r *BlockchainReactor) Stop() {
+    ...
+    r.msgs <- stop
+    ...
+}
+...
+
+func (r *BlockchainReactor) AddPeer(peer p2p.Peer) {
+    ...
+    r.msgs <- bcAddPeerEv{peer.ID}
+    ...
+}
+
+```
+
+## IO handling
+An io handling routine within the reactor will isolate peer communication.
+
+```go
+func (r *BlockchainReacor) ioRoutine(chan ioMsgs, ...) {
+    ...
+    for {
+        msg := <-ioMsgs
+        switch msg := msg.(type) {
+            case scBlockRequestMessage:
+                r.sendBlockToPeer(...)
+            case scStatusRequestMessage
+                r.sendStatusResponseToPeer(...)
+            case bcPeerError
+                r.Swtich.StopPeerForError(msg.src)
+                ...
+            ...
+            case bcFinished
+                break;
+        }
+    }
+}
+
+```
+### Processor internals 
+
+The processor will be responsible for validating and processing blocks.
+The Processor will maintain an internal cursor `height` of the last
+processed block. As a set of unordered blocks arrive, the the Processor
+will check if it has `height+1` nessary to process the next block.
+
+```go
+type Proccesor struct {
+    height ...
+    state ...
+    blocks [height]*Block
+    peers[height]PeerID
+    lastTouch timestamp
+}
+
+func (proc *Processor) handleBlockResponse(peerID, block, time) {
+    lastTouch = time
+    if block.height < height {
+        // skip
+    } else if blocks[block.height] {
+        return errDuplicateBlock{}
+    } else  {
+        blocks[block.height] = block
+    }
+
+    if blocks[height] && blocks[height+1] {
+        ... = processBlock(blocks[height], blocks[height+1])
+         bcR.store.SaveBlock(first, firstParts, second.LastCommit)
+        ... = proc.state.Validators.VerifyCommit
+        delete blocks[height]
+        height++
+    }
+    return pcBlockProcessed{height}
+}
+
+func (proc *Processor) handleRemovePeer(peerID) {
+    events = []
+    // Delete all unprocessed blocks from peerID
+    for i = height; i < len(blocks); i++ {
+        if peers[i] == peerID {
+            events = append(events, pcBlockReschedule{height})
+            delete block[height]
+        }
+    }
+    return events
+}
+
```

Comment:
From my understanding, The processor is responsible for removing/rescheduling all unprocessed block received from an errored peer. Is that true?

Reply:
IMHO we need to clarify this. Looks like both the processor and the scheduler maintain some information about the fast sync peers and blocks. They may both detect errors and trigger peer removal and eventually the rescheduling of requests. Crossing messages may cause their view of peers and blocks to get out of sync. At this point the scheduler struct is not filled in, also how the actual scheduling works is missing. I think we need more details on the scheduler and then make the analysis.

Reply:
> The processor is responsible for removing/rescheduling all unprocessed block received from an errored peer. Is that true?
One way to think about Processor is that it should order, verify and execute blocks. If it fails by doing this for some block it will notify scheduler (and io) about error so they can do corresponding actions.